omx/gstomxvideofilter.c

```diff
@@ -0,0 +1,2398 @@
+{
+  g_slice_free (GstVideoInfo, info);
+}
+
+/* called to add a GstVideoCodecFrame to the pad pending list given
+ * in @value*/
+static void
+add_frame_to_pad_array (gpointer key, gpointer value, gpointer user_data)
+{
+  GstVideoCodecFrame *frame = (GstVideoCodecFrame *) user_data;
+  gst_video_codec_frame_ref (frame);
+  g_ptr_array_add ((GPtrArray *) value, frame);
+}
+
+/* called to free @value pad frame pending list, unref each
+ * frame before freeing */
+static void
+free_frame_array (gpointer key, gpointer value, gpointer user_data)
+{
+  g_ptr_array_foreach ((GPtrArray *) value, (GFunc) gst_video_codec_frame_unref,
+      NULL);
+  g_ptr_array_free ((GPtrArray *) value, TRUE);
+}
+
+
+static GstElementClass *parent_class = NULL;
+
+static void gst_omx_video_filter_class_init (GstOMXVideoFilterClass * klass);
+static void gst_omx_video_filter_init (GstOMXVideoFilter * self,
+    GstOMXVideoFilterClass * klass);
+
+static void gst_omx_video_filter_finalize (GObject * object);
+static void gst_omx_video_filter_set_property (GObject * object, guint prop_id,
+    const GValue * value, GParamSpec * pspec);
+static void gst_omx_video_filter_get_property (GObject * object, guint prop_id,
+    GValue * value, GParamSpec * pspec);
+
+static gboolean gst_omx_video_filter_set_caps (GstOMXVideoFilter * self,
+    GstCaps * incaps);
+static GstCaps *gst_omx_video_filter_sink_get_caps (GstOMXVideoFilter * self,
+    GstCaps * filter);
+static GstStateChangeReturn gst_omx_video_filter_change_state (GstElement *
+    element, GstStateChange transition);
+static gboolean gst_omx_video_filter_sink_event (GstPad * pad,
+    GstObject * parent, GstEvent * event);
+static gboolean gst_omx_video_filter_sink_query (GstPad * pad,
+    GstObject * parent, GstQuery * query);
+static GstFlowReturn gst_omx_video_filter_chain (GstPad * pad,
+    GstObject * parent, GstBuffer * buf);
+
+static GstFlowReturn gst_omx_video_filter_drain (GstOMXVideoFilter * self);
+
+static GstCaps *gst_omx_video_filter_default_transform_caps (GstOMXVideoFilter *
+    self, GstPadDirection direction, GstPad * srcpad, GstCaps * caps,
+    GstCaps * filter);
+static GstCaps *gst_omx_video_filter_default_fixate_caps (GstOMXVideoFilter *
+    self, GstPad * srcpad, GstCaps * sinkcaps, GstCaps * srccaps);
+
+static gboolean gst_omx_video_filter_stop (GstOMXVideoFilter * self);
+static gboolean gst_omx_video_filter_shutdown (GstOMXVideoFilter * self);
+
+gint gst_omx_video_filter_get_buffer_size (GstVideoFormat format, gint stride,
+    gint height);
+OMX_COLOR_FORMATTYPE gst_omx_video_filter_get_color_format (GstVideoFormat
+    format);
+
+enum
+{
+  PROP_0,
+  PROP_ALWAYS_COPY,
+  PROP_OUTPUT_BUFFERS,
+  PROP_INPUT_BUFFERS
+};
+
+#define GST_OMX_VIDEO_FILTER_ALWAYS_COPY_DEFAULT FALSE
+#define GST_OMX_VIDEO_FILTER_OUTPUT_BUFFERS_DEFAULT 6
+#define GST_OMX_VIDEO_FILTER_INPUT_BUFFERS_DEFAULT 6
+
+/* we can't use G_DEFINE_ABSTRACT_TYPE because we need the klass in the _init
```

Comment:
I'm pretty sure we still can use `G_DEFINE_TYPE` and get the class

Reply:
I tried that but in the _init method I couldn't get access to the kclass, in that case the method definition can be just something like gst_omx_video_filter_init (GstOMXVideoFilter * self)